tests/ValidatorTest.php

```diff
@@ -283,9 +283,9 @@ public function testDiverseTypes($dataSet): void
     public function testNullAsDataSet(): void
     {
         $validator = new Validator();
-        $result = $validator->validate(null, ['property' => [new Compare(null)]]);
```

Comment:
We can keep this test and add new.

Reply:
It no longer can be kept after these changes:

https://github.com/yiisoft/validator/blob/7dfab7f12b01300841c7d5fae035a5d1682672c2/src/Rule/AbstractCompare.php#L132-L134

Would be the same as:

https://github.com/yiisoft/validator/blob/67313f798abf79801e123aa1262d88ff5b5d09be/tests/Rule/CompareTest.php#L23-L29

Reply:
A more in-depth explanation and update.

Actually, initially I found that this check was present in shortcuts but missing in base rule:

https://github.com/yiisoft/validator/blob/7dfab7f12b01300841c7d5fae035a5d1682672c2/src/Rule/AbstractCompare.php#L132-L134

This implies 2 problems:

- Copy paste.
- Inconsistency in logic because both `Compare` and shortcuts are available for usage.

I moved it to the base rule, but this revealed another problem - not being able to specify `null` as `targetValue` (previously it was the case too, but only for shortcuts). Then i figured it out that this check is not needed at all. Precedence of `targetValue` is handled and documented.

I reverted according changes, so, at the end, this test can be kept.